### ISSUE TITLE
Removing Domains and Identities does not remove them from the filesystem

### ISSUE THREAD

**Issue by MrGeneration** [modoboa/modoboa#2157]
# Impacted versions

* OS Type: Debian
* OS Version: Buster / 10
* Database Type: MySQL
* Database version: 15.1
* Modoboa: 1.17.0 (I also saw the issue on earlier versions (1.15.x and 1.16.x)
* installer used: Yes
* Webserver: Nginx

# Steps to reproduce

* create a domain
* add a new account to that domain, retrieve a mail or two
* delete account
* delete domain
* wait a minute, days - doesn't matter
* check your vmail directory for the domain and account

# Current behavior

Modoboa claims that it has removed a user account (and domain).
It no longer appears within the web interface, however, you can still find data for account and domain within the vmail directory on the server.

# Expected behavior

Modoboa removes account- and domain folders as it claims it does.
Right now this behavior means manual work on the console for cleaning up.

# Video/Screenshot link (optional)
--

**Comment by tonioo** [comment]
@MrGeneration This is handled by the manage_mailbox_operations cron script and it will work only if Modoboa is installed on the same server than dovecot.

**Comment by MrGeneration** [comment]
Yes.
I did run that command next to the crontab just to be sure that it's not working.
All services and modoboa are running on the exact same maschine.

Are specific permissions required for the tasks to run as expected?
I also noticed certificate (also for it's key) permission errors if the certificate files referenced in the dovecot configuration would not be world readable.

**Comment by tonioo** [comment]
This cron script must be run as vmail user and please check that the feature is activated in the administration panel (look at the administration tab in the parameters section).

About the error, can you paste a trace?

**Comment by MrGeneration** [comment]
As for the certificate based error I'd provide another issue as soon as I did find time to undo my change and reproduce it.

Modoboa did actually log the following line during my deletion attempt:

`Delete doveadm(vmail): Error: net_connect_unix(/var/run/dovecot/stats-writer) failed: Permission denied /srv/vmail/example.com/support succeed`

This is a standard installation which why this is odd.
I'll also check on my behalf later on. I just wanted to provide this feedback for now so it's not lost.

**Comment by MrGeneration** [comment]
So I set `Debug: True` within `setting.py` and restarted `uwsgi.service` before trying this time.
There are no errors logged, I did wait for over an hours and even run `handle_mailbox_operations` manually to be extra sure.

No errors are logged.
The directories are not gone.

Permission for said directories and files are `700` and owned by `vmail:vmail`.

If I understood wrong on how to enable debugging, please let me know how to do it correctly.

**Comment by tonioo** [comment]
@MrGeneration Can you check if you have records in the admin_mailboxoperation table?

**Comment by MrGeneration** [comment]
Sure! The table is empty.

**Comment by tonioo** [comment]
Ok, so it means the script won't have anything to do. When you delete something, an order is created in this table and is removed once the job is done. Can you perhaps disable the cron, delete something and check if a record is created in this table ?

**Comment by MrGeneration** [comment]
That's interesting. 
So I disabled the cron and then checked the table.

### What I did:
* deleted the mailbox (tick for "do not remove from storage" was unselected)
* deleted the domain (tick for "do not remove from storage" was unselected)

My `admin_mailboxoperation` table now contains one entry:
```
== Table structure for table admin_mailboxoperation

|------
|Column|Type|Null|Default
|------
|//**id**//|int(11)|No|
|type|varchar(20)|No|
|argument|longtext|No|
|mailbox_id|int(11)|Yes|NULL
== Dumping data for table admin_mailboxoperation

|12|delete|doveadm(vmail): Error: net_connect_unix(/var/run/dovecot/stats-writer) failed: Permission denied
/srv/vmail/example.com/customer|NULL
```

This lead me to checking for the stats-writer permissions (the UI did not show that error any more):
```
[...]
srw-rw---- 1 root dovecot 0 Mar 18 11:51 /var/run/dovecot/stats-writer
[...]
```

So I then re-added the example domain along with a test user and ensures the folders were written. 
I added `vmail` user to `dovecot` user manually by using `# usermod -a -G dovecot vmail`.

I repeated above removal steps. This time the result looks much better:
```
== Table structure for table admin_mailboxoperation

|------
|Column|Type|Null|Default
|------
|//**id**//|int(11)|No|
|type|varchar(20)|No|
|argument|longtext|No|
|mailbox_id|int(11)|Yes|NULL
== Dumping data for table admin_mailboxoperation

|13|delete|/srv/vmail/example.com/user|NULL
```

Even though I would expect an entry for the domain folder as well. 
Any way. I run `manage.py handle_mailbox_operations` manually and was re-visitting my folder structure. 
Now the removal process does work as expected.

----

This leads me to the question if I turned left on a step where I shouldn't have. 
I followed migration and installation documentation and initially even used the installer. 
For me the issue is now solved, however, we should ensure that the installer does.

After re-visitting the documentation just now I found this warning:
```
The cron script must be executed by the system user owning the mailboxes.
```

which makes perfect sense. However - the error above is thrown before the mailbox operation can even start.
It seems like the operation script "ignores" the faulty entry and removes it silently.

Currently my time resources and python knowledge is not good enough for hunting this issue on my own, sorry.

**Comment by tonioo** [comment]
The system user running modoboa (generally modoboa) must have access to the doveadm command. I fixed a bug regarding this a few weeks ago but it has not been released yet (https://github.com/modoboa/modoboa/commit/1827bce52d8002060b68ec24752080f5af7a2a76).

**Comment by kiloDE** [comment]
> The system user running modoboa (generally modoboa) must have access to the doveadm command. I fixed a bug regarding this a few weeks ago but it has not been released yet ([1827bce](https://github.com/modoboa/modoboa/commit/1827bce52d8002060b68ec24752080f5af7a2a76)).

**Thus "Bug" here? ...** `www-data : user NOT in sudoers ; TTY=unknown ; PWD=/srv/modoboa/instance ; USER=vmail ; COMMAND=/usr/bin/doveadm user -f home <email>`

I can't fix it via sudoers File. Until yesterday everything ran fine. Now it's fully broken in some areas. That sucks :D 
I compared the patch you made to the local file version, in py3.8 and recently updated the git repo is more in the future than the local version. But I don't see, except for the sudoers file solution, where to fix that.

Did you changed something around the doveadm_cmd Function? Except the line itself and cleaning it a bit up?

**Comment by gmartinez** [comment]
Same here, on Ubuntu 20.04LTS;

Tried both deleting via web UI and API, same result.

Record in table/column admin_mailboxoperation/argument, reads;

_doveconf: Warning: NOTE: You can get a new clean config file with: doveconf -Pn > dovecot-new.conf
doveconf: Warning: Obsolete setting in /etc/dovecot/conf.d/10-ssl.conf:45: ssl_protocols has been replaced by ssl_min_protocol
doveadm(vmail): Error: net_connect_unix(/var/run/dovecot/stats-writer) failed: Permission denied
/srv/vmail/{domain}/johndoe_

I guess the warnings can be ignore and probably even fixed by changing the indicated deprecated parameter, but the "Permission denied" is why the folder /srv/vmail/{domain}/johndoe does not get deleted.

I am running a fresh installation (v1.17.0 dated 11-05-20) so I assume the fix @tonioo mentioned at commit 1827bce is still not part of it.

I guess my question to both @tonioo and @MrGeneration is whether the fix proposed (adding user vmail to group dovecot) is ok/safe, at least in the meantime, or there is something more "official" from the aforementioned commit that can be manually applied (I am not python savvy but I can try :).

Thanks in advance.
Gonzalo.

**PS;** The corresponding API endpoint might benefit from a parameter so one could enforce the "Do not delete mailbox directory" option available in the web UI when deleting an account.

**Comment by gmartinez** [comment]
@MrGeneration seems to have done it for me as well, although 2nd point in your list was already present in my system inside as a /etc/sudoers.d/modoboa file, adding dovecot to vmail group seem to have done the trick. Thanks !!!

**Comment by virtuaris** [comment]
This issue is still present. Installation of `modoboa` (2.0.1) via `modoboa-installer` (HEAD 9ff989a) on vanilla Debian 10 (Buster) minimal.
One need to put user `vmail` into group `dovecot` as stated above by @MrGeneration or the mailbox filesystem commands (delete, rename) will fail silently.
This is an essential bug IMHO but easy to fix in the installer. Why are new users still being bothered with this?

**Comment by ksaadDE** [comment]
> This is an essential bug IMHO but easy to fix in the installer. Why are new users still being bothered with this?

can you create a fix and file a PR?

**Comment by virtuaris** [comment]
> Modoboa claims that it has removed a user account (and domain).
> It no longer appears within the web interface, however, you can still find data for account and domain within the vmail directory on the server.

Did as requested: https://github.com/modoboa/modoboa-installer/pull/427.

Tested this on Debian 10 vanilla and it works. Group `dovecot` is hardcoded which may not be perfect but the current Dovecot docs state that the `dovecot` system user should be member of a group with the same name. As long as the supported Linux distributions adhere to this guideline, everything should work.
